src/poller/mod.rs

```diff
@@ -0,0 +1,184 @@
+use crate::{
+    audit::log_sync,
+    common::get_unix_time_seconds,
+    db::{delete, get_all},
+    erc20::ERC20Token,
+    gateway::PaymentGateway,
+    transfers::{errors::TransferError, gas_transfers::transfer_gas_to_treasury},
+    types::Invoice,
+};
+
+use alloy::{
+    primitives::Uint,
+    providers::{Provider, RootProvider},
+    rpc::types::eth::TransactionReceipt,
+    transports::http::Http,
+};
+use reqwest::Client;
+use sled::Tree;
+
+/// Checks if a specific token of a specific amount has been received
+/// at a certain address.
+async fn check_if_token_received(
+    token: ERC20Token,
+    invoice: Invoice,
+) -> Result<bool, alloy::contract::Error> {
+    let balance_of_recipient = token.get_balance(invoice.to).await?;
+    if balance_of_recipient.ge(&invoice.amount) {
+        return Ok(true);
+    }
+    Ok(false)
+}
+
+/// Retrieves the gas token balance of the specified address on the specified web3 instance
+async fn get_native_balance(
+    provider: RootProvider<Http<Client>>,
+    address: String,
+) -> Result<Uint<256, 4>, alloy::contract::Error> {
+    Ok(provider
+        .get_balance(address.parse().unwrap(), alloy::eips::BlockId::latest())
+        .await?)
+}
+
+/// Used to check if the invoice recipient has received enough money to cover the invoice
+async fn check_if_native_received(
+    provider: RootProvider<Http<Client>>,
+    invoice: Invoice,
+) -> Result<bool, alloy::contract::Error> {
+    let balance_of_recipient = get_native_balance(provider, invoice.to).await?;
+    if balance_of_recipient.ge(&invoice.amount) {
+        return Ok(true);
+    }
+    Ok(false)
+}
+
+/// A function that branches control flow depending on the invoice shall
+/// be paid by an ERC20-compatible token or the native gas token on the network
+async fn check_and_process(provider: RootProvider<Http<Client>>, invoice: Invoice) -> bool {
```

Comment:
accept by ref, it reduces unnecessary clones

Reply:
https://github.com/starkbamse/acceptevm/pull/1/commits/